keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1046 @@
+	NodeName           string `asn1:"utf8"`
+}
+```
+
+Certificates and CertificateRequests that embed this extension do not need to include any data in the standard X.509 Name or Subject Alternate Name fields.  The CertificateRequests generated by Kubelet will not, nor will the certificates issued by `kubernetes.io/kube-apiserver-client-pod`.
+
+Utility functions for embedding and extracting this extension from X.509
+Certificate and CertificateRequest objects are defined in
+`k8s.io/component-helpers/kubernetesx509`.  They are intended to be used both by
+core Kubernetes components and third-party software that needs to interact with
+these certificates.
+
+### PodIdentity Extension Node Restriction
+
+The noderestriction admission plugin will check the consistency of PodIdentity extensions in CertificateSigningRequest objects created by `system:node:` identities.  The creation request will only be admitted if the PodIdentity extension refers to a pod that is currently running on the node that requested the certificate.
+
+This upholds the Kubernetes node isolation guarantee by ensuring that
+compromising a single node only results in compromises of workloads that have
+actually been scheduled onto that node.
+
+<<[UNRESOLVED]>>
+Is it necessary or desirable to forbid non-Kubelet identities from creating CertificateSigningRequest objects that embed a PodIdentity extension?
+<<[/UNRESOLVED]>>
+
+During Alpha, this capability is behind the PodIdentityX509Support feature gate.
+
+### PodCertificate Projected Volume Sources
+
+The PodCertificate projected volume source instructs Kubelet to inject a
+credential bundle consisting of a private key and certificate chain into the
+container fileystem.  Kubelet handles generating the key, requesting a
+certificate from the named signer, and renewing the certificate before it
+expires.
+
+Kubelet supports only a few key generation strategies, each named by a string
+that encapsulates both the key type and any parameters necessary (for example,
+RSA modulus size).  The intention is to offer a a tasting menu of reasonable key
+choices, rather than offering a flexibility to a wide variety of parameters.
+Key types supported by Kubelet are "RSA2048", "RSA3072", "RSA4096", "ECDSAP256",
+and "ECDSAP384".  If no key type is specified, Kubelet defaults to "ECDSAP256".
+
+<<[UNRESOLVED]>>
+Should RSA2048 be offered?  I'm inclined to say no.  Are there any other key types that should be offered?
+<<[/UNRESOLVED]>>
+
+Once the key is generated, Kubelet creates a CertificateSigningRequest that
+contains an X.509 CertificateRequest with no data in the Name or
+SubjectAlternateName fields, but instead embeds an X.509 PodIdentity extension
+with the details of the pod that mounted the volume.  The signer is intended to treat this CertificateRequest merely as an indication that it should issue a certificate for the named pod, rather than trying to issue a certificate that follows the structure of the CertificateRequest.
```

Comment:
```suggestion
with the details of the pod that mounted the volume. 
The signer is intended to treat this certificate request merely as an indication that it
should issue a certificate for the named pod, rather than trying to issue a certificate
that follows the structure of the CertificateRequest.
```
Also, can we explain why there's that mismatch between CSR and issued certificate?

Reply:
I'm a bit confused about what needs clarification; a Kubernetes signer implementation has never been required to pay any attention to the certificate request embedded in a CertificateSigningRequest object.  It's totally free to to issue whatever certificate it likes.

Reply:
We haven't documented the [lack of] link between CertificateSigningRequest and certificates. https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#certificate-signing-requests implies that there is a link.

I know that the way these things work allow issuing an semi-related certificate that uses the same public key, but that's not the usual way people would interpret the name. So it's important to make it clear when “certificate signing request” actually means: “please sign some other certificate, not this one”.